### ISSUE TITLE
Type annotation on global mutable variable is ignored

### ISSUE THREAD

**Issue by matejcik** [microsoft/pyright#7780]
**Describe the bug**
My module has a `GLOBAL` symbol that can be modified from outside.
Pyright does not like that.

Most simplified example:
```
GLOBAL: int | None = None

def bar() -> None:
    reveal_type(GLOBAL)
    assert GLOBAL is not None
    reveal_type(GLOBAL)
```
output:
```
/home/matejcik/projekty/prace/trezor-firmware/core/repro.py
  /home/matejcik/projekty/prace/trezor-firmware/core/repro.py:4:17 - information: Type of "GLOBAL" is "None"
  /home/matejcik/projekty/prace/trezor-firmware/core/repro.py:6:17 - information: Type of "GLOBAL" is "Never"
```

I would expect that if I `assert GLOBAL is not None`, its type should be `int`.

Surprisingly this works for simple types but causes a problem in a more complex scenario:
```
class Foo:
    async def foo(self) -> int:
        ...


GLOBAL: Foo | None = None

async def bar() -> int:
    assert GLOBAL is not None
    return await GLOBAL.foo()
```
output:
```
/home/matejcik/projekty/prace/trezor-firmware/core/repro.py
  /home/matejcik/projekty/prace/trezor-firmware/core/repro.py:10:18 - error: "Never" is not awaitable (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
both
pyright 1.1.360

**Comment by matejcik** [comment]
even adding `assert isinstance(GLOBAL, Foo)` will not make the error go away, the inferred type is still `Never`

**Comment by erictraut** [comment]
Pyright is working as intended here, but I can understand why this looks like a bug.

Pyright includes some special-case handling of captured variables. You can read about it [here](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=narrowing-for-captured-variables).

The problem is that if I modify this behavior to exempt global (module-level) variables from this logic, it will create false positive errors in other common cases.

```python
GLOBAL: int | None = 1

def foo():
    return GLOBAL + 1
```

It will also mask some errors that are currently caught by pyright, like this one:

```python
if 1 + 1 > 3:
    y = 0

def func1(self):
    print(y)
```

I don't know if this is a good tradeoff.

You can work around pyright's behavior by moving the global definition below its use — or by reassigning it somewhere else within the module, even if it's conditional:

```python
if False:
    GLOBAL = None
```

I'll note that your code sample is using a variable with an all-caps name, which by convention implies that it's a constant (implicitly `Final`) and should not be modified after its initial assignment. Under that circumstance, the current behavior seems justified. However, if you were to change its name to `Global` or `my_global`, then this argument no longer applies.

Changing this behavior is straightforward, so I've made the change on a branch. As I feared, this change generates new [false positives and negatives](https://github.com/microsoft/pyright/pull/7793#issuecomment-2081238487) in code bases that are currently using pyright for type checking.

I'm going to leave this issue open for a little while to give it more thought and give you a chance to respond, but my current inclination is to leave the current behavior in place and close the issue as "won't fix, as designed".

**Comment by matejcik** [comment]
> ```python
> if False:
>     GLOBAL = None
> ```

for the record, this doesn't work because pyright is smart enough to figure out that this doesn't execute. changing to `if TYPE_CHECKING:` works fine.

> ```python
> GLOBAL: int | None = 1
> 
> def foo():
>     return GLOBAL + 1
> ```

This one does not look like a false positive to me. Someone took the pain to declare that the variable is allowed to be `None`. The declaration is not explicitly `Final`, and a different module _would not trigger an error_ by assigning `thismod.GLOBAL = None`

(acknowledging that it might be reasonable to treat all-upper names as `Final`, so from now on let's assume that the names are all lowercase)

> ```python
> if 1 + 1 > 3:
>     y = 0
> 
> def func1(self):
>     print(y)
> ```

For the record, this does not cause an error on my version, but the following does:
```
if 2 > 3:
    y = "helo"

def func1():
    print(y + "!")
```
(are ints special? `reveal_type` says the correct thing in your case, but will not fail the typecheck, not even if i say `y + 1`)

This also seems to be a bug in the implementation of the fix. The code above should still fail after the fix, the type of `y` should still be `str | Unbound`.

[Similar looking code](https://github.com/DataDog/dd-trace-py/blob/main/ddtrace/vendor/monotonic/__init__.py#L99C1-L130C59) is kind-of a false positive ...... buuuuut again, `GetTickCount64` here is inferred to something like `None | Callable`, and external modules are _technically_ allowed to change it externally back to `None` with no type error, absent an explicit `Final` type.

---

The way I see it, this is a behavior that avoids false positives in commonly used code, that would require explicit `Final` annotation to fix, so from a practicality standpoint you might not want to just break all that code out there -- although it's _technically_ incorrect.

But on the other hand, there is no semantically correct way to fix the false negative in my original case. Like, sure, if there is some sort of `Nonfinal[Foo | None]` that I could put on the variable, I'll just go do that. But "make the typechecker think that something modifes the variable" does not seem like the way to go.

How about applying a heuristic: if there is an explicit type annotation on the global variable, just keep it?
(i also vaguely expected that if I annotated the usage with `global GLOBAL`, it would help -- although there is probably no rule saying that it should)

**Comment by erictraut** [comment]
> For the record, this does not cause an error on my version

I suspect that you have disabled the `reportPossiblyUnboundVariable` diagnostic check. With a default configuration, pyright will report that `y` is possibly unbound in [this code](https://pyright-play.net/?code=JYMwBAjGDUlgfGAzALgLACgzbATzALxgAMmZGAJgKbggCuAdgMYQAUAzlQDYgCU6WHAAcATsAYAXVrl6YgA).

> there is no semantically correct way to fix the false negative in my original case

Yeah, that occurred to me too, and I think it's the best argument in favor of changing the behavior, but I'm not 100% convinced yet. I'll think about it more.

**Comment by erictraut** [comment]
I've decided to move ahead with this change because I agree that it's technically correct. It is going to result in some false positives though, and it may cause some pain for some pyright users. 

The change will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.361, which I just published.
